File: lr/src/questions/q_0130_surrounded_regions.rs
#[allow(clippy::ptr_arg)]
pub fn solve(board: &mut Vec<Vec<char>>) {
    let (rows, cols) = (board.len() as i32, board[0].len() as i32);

    fn dfs(rows: i32, cols: i32, board: &mut [Vec<char>], row: i32, col: i32) {
        if row < 0
            || row >= rows
            || col < 0
            || col >= cols
            || board[row as usize][col as usize] != 'O'
        {
            return;
        }

        board[row as usize][col as usize] = 'F';
        dfs(rows, cols, board, row, col + 1);
        dfs(rows, cols, board, row, col - 1);
        dfs(rows, cols, board, row + 1, col);
        dfs(rows, cols, board, row - 1, col);
    }
    for row in 0..rows {
        for col in 0..cols {
            if board[row as usize][col as usize] == 'O'
                && (row == 0 || col == 0 || row == rows - 1 || col == cols - 1)
            {
                dfs(rows, cols, board, row, col)
            }
        }
    }

    for row in 0..rows {
        for col in 0..cols {
            if board[row as usize][col as usize] == 'O' {
                board[row as usize][col as usize] = 'X';
            }
        }
    }

    for row in 0..rows {
        for col in 0..cols {
            if board[row as usize][col as usize] == 'F' {
                board[row as usize][col as usize] = 'O';
            }
        }
    }
}

#[test]
fn test_solve() {
    let mut board = vec![
        vec!['X', 'X', 'X', 'X'],
        vec!['X', 'O', 'O', 'X'],
        vec!['X', 'X', 'O', 'X'],
        vec!['X', 'O', 'X', 'X'],
    ];
    solve(&mut board);
    assert_eq!(
        board,
        vec![
            vec!['X', 'X', 'X', 'X'],
            vec!['X', 'X', 'X', 'X'],
            vec!['X', 'X', 'X', 'X'],
            vec!['X', 'O', 'X', 'X'],
        ]
    );
}

File: lg/src/questions/q_0130_surrounded_regions.go
package questions

func solve(board [][]byte) {
	rows, cols := len(board), len(board[0])
	directions := [][]int{
		{1, 0},
		{-1, 0},
		{0, 1},
		{0, -1},
	}

	var markBorderConnected func(r, c int)
	markBorderConnected = func(r, c int) {
		if r < 0 || r >= rows || c < 0 || c >= cols || board[r][c] != 'O' {
			return
		}
		board[r][c] = 'B'
		for _, direction := range directions {
			markBorderConnected(r+direction[0], c+direction[1])
		}
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if (r == 0 || r == rows-1 || c == 0 || c == cols-1) && board[r][c] == 'O' {
				markBorderConnected(r, c)
			}
		}
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if board[r][c] == 'O' {
				board[r][c] = 'X'
			}
		}
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if board[r][c] == 'B' {
				board[r][c] = 'O'
			}
		}
	}
}
